Control flow AST followup fixes

crisbeto opened the pull request:
Includes the following fixes to control flow AST:

### refactor(compiler): don't allow as expressions in else if blocks
Based on some discussions, these changes remove the ability to have an `as` expression on an `else if` block.

### refactor(compiler): parse for loop track as an expression
Adds some logic to store the `track` parameter of a `for` loop block as an expression AST instead of a string.

### refactor(compiler): parse let parameters in for loop
Adds the logic to parse `let` parameters of a `for` loop block which was missed in #51299.

crisbeto committed:
refactor(compiler): don't allow as expressions in else if blocks

Based on some discussions, these changes remove the ability to have an `as` expression on an `else if` block.

crisbeto committed:
refactor(compiler): parse for loop track as an expression

Adds some logic to store the `track` parameter of a `for` loop block as an expression AST instead of a string.

alxhub commented on packages/compiler/src/render3/r3_ast.ts:
```diff
@@ -257,13 +257,20 @@ export class SwitchBlockCase implements Node {
+  $first?: string;
+  $last?: string;
+  $even?: string;
+  $odd?: string;
```
Do we allow `$count` as well?

crisbeto replied:
I didn't see `$count` mentioned in the RFC so I didn't include it here. For reference https://github.com/angular/angular/discussions/50719

crisbeto commented:
@alxhub I've reworked the changes to use the `d` flag on the regex as we discussed over DM.

crisbeto committed:
refactor(compiler): parse let parameters in for loop

Adds the logic to parse `let` parameters of a `for` loop block which was missed in #51299.

AndrewKushnir commented:
This PR was merged into the repository by commit 36663e6ef6d7f9b7e64cd3c81e1dd2b2cbaf7723.
